### ISSUE TITLE
How to have access to simple python APIs

### ISSUE THREAD

**Issue by lin-lin** [malmaud/TensorFlow.jl#290]
Thanks for the great package.  I am a new user of Julia. While I understand it is quite tedious to implement all python APIs especially if some of them are trivial, I wonder whether there is some easy way to get access to some of the Tensorflow functions implemented natively in Python. 

For instance, tf.train.exponential_decay is not implemented in Tensorflow.jl, but it would be nice to use it without digging into the python code.

**Comment by malmaud** [comment]
There isn't a great way of doing this. You can reference the Python TensorFlow functions with `@TensorFlow.py_call py_tf[:train]...` but it's kind of complicated and not robust to mix-and-matching with the Julia TF API.

**Comment by lin-lin** [comment]
The problem is some functions and variables (like global_step) are implicitly updated, and the mix-and-match procedure presumably would not work. What is the plan then for such APIs? I found the Julia wrapper works like a charm for the test examples, but when going a bit deeper, such API related issues make me feel that I should go back to Python to be compatible with other people's results.

**Comment by oxinabox** [comment]
For things like Expodential Decay, I think it is fully reasonable to just add equivlent functions implemented in julia -- none of this stuff is hard see: https://github.com/malmaud/TensorFlow.jl/blob/master/src/train.jl.
In the case of Expodential Decay in particular I don't find it overless pressing, as most new research I am seeing (/doing) is using Adam anyway, which is not normally used with learning rate decay (in a way it does it implictly).
Good to have eventually for reproducing old results, for sure though.


All *core* functionality of tensorflow is automatically implemented in the Ops submodule.
Here using the definition of Core being that it was implemented in tensorflow backend rather than the client libraries.
However an issue is that a surprising amount of tensorflow functionality in in the various clients.
(and may be present in some, eg python, but not in others eg C++).
I believe the tensorflow team is moving more stuff into the backend, which means we will get it for free.

But really I think it is viable just to implement these things, like every other client has to.

> The problem is some functions and variables (like global_step) are implicitly updated, and the mix-and-match procedure presumably would not work. 

I don't think so. They are not "implicitly updated"
You [have to pass them in, no](https://www.tensorflow.org/api_docs/python/tf/train/exponential_decay)?
`global_step` is a real node in the graph, you should be be able to pass it in (with some fiddling perhaps).

**Comment by lin-lin** [comment]
@oxinabox It would be fine for me to implement a few APIs myself if it is simple, but I have not figured out how to implement it according to the examples in train.jl yet. Although exponential_decay is a particularly simple example, what I would like to do is to call the python interface, and not to translate the python implementation to the julia one.  For example, after looking into the implementation of "gradients" function in TensorFlow.jl (which is needed by compute_gradients in train.jl), it seems that I need to use some syntax like 

node_protos, grad_names = fetch(@py_proc py_gradients($graph_proto, $x_names, $y_names, $grad_y_names))

for such purpose. However, I am rather new to Julia and not sure how to do this for other functions. If not too much trouble do you mind showing me how to do this, say for exponential_decay?

By "implicit updating" I mean that some variables are updated by TensorFlow. For instance, global_step is updated by the advance_step macro. However, I was not sure whether the mix-and-match would get everything right, and I felt more comfortable if this can be done consistently.  Thanks.

**Comment by malmaud** [comment]
I don't think we intend to implement stuff in this package to make it a supported workflow to directly call Python tensorflow. Ideally someone could port the relevant functionality to native Julia, or you can bypass this package entirely and just use PyCall with the Python tensorflow.
